### PR TITLE
fix: fall back to FTS when search embeddings stall

### PR DESCRIPTION
**Pull request by EtanHey**
## Summary
- Time-bound MCP brain_search query embedding via BRAINLAYER_EMBED_TIMEOUT_MS and fall back to FTS-only on timeout, embed errors, or an already stalled embedder.
- Allow hybrid_search to accept query_embedding=None and skip vector work while keeping the FTS/RRF result path alive.
- Time-bound prompt injection hybrid search and fall back to FTS-only; operational/no-result injection event noise remains suppressed.
- Apply the same timeout/FTS fallback guard to the live ~/.claude prompt hook.

## Tests
- RED first: slow/raising embedder tests failed before implementation.
- pytest tests/test_search_handler.py -k "falls_back_to_fts_when_embedding_times_out or falls_back_to_fts_when_embedding_raises or uses_hybrid_when_embedding_is_fast or parallel_embedding or empty_fts_fallback" -q
- pytest tests/test_adaptive_injection.py -k "fallback_to_fts_only or slow_hybrid or timed_out_hybrid or fast_hybrid" -q
- python3 -m py_compile src/brainlayer/mcp/search_handler.py src/brainlayer/search_repo.py hooks/brainlayer-prompt-search.py /Users/etanheyman/.claude/hooks/brainlayer-prompt-search.py
- pytest tests/test_search_handler.py tests/test_adaptive_injection.py tests/test_hybrid_search.py tests/test_conditional_hooks.py tests/test_search_filter_params.py tests/test_search_profile.py -q
- ruff check src/brainlayer/mcp/search_handler.py src/brainlayer/search_repo.py hooks/brainlayer-prompt-search.py tests/test_adaptive_injection.py tests/test_hybrid_search.py tests/test_search_handler.py
- ruff format --check src/brainlayer/mcp/search_handler.py src/brainlayer/search_repo.py hooks/brainlayer-prompt-search.py tests/test_adaptive_injection.py tests/test_hybrid_search.py tests/test_search_handler.py
- BRAINLAYER_PREPUSH=1 scripts/run_tests.sh via pre-push hook: 3019 passed, 9 skipped, 61 deselected, 1 xfailed; MCP registration 3 passed; isolated eval/hook routing 40 passed; bun 1 passed; regression shell passed.

Supersedes #523 because GitHub did not synchronize #523 after the follow-up commit even though the branch ref advanced.

---

> [!NOTE]
> **Medium Risk**
> Changes core brain_search and prompt-hook retrieval concurrency and response shape (`search_mode`, `fallback_reason`); lock bugs could degrade all searches to FTS-only, but behavior is covered by new timeout/busy/cancel tests.
> 
> **Overview**
> Prevents **stacked embedding work** when hybrid search or MCP `brain_search` is slow or times out. A process-wide lock means a second concurrent request skips vector embedding and uses **FTS-only** immediately (`embed_busy`) instead of starting another embedder call.
> 
> The MCP `_search` path wraps executor-based embedding with **lock acquire/release** on timeout, errors, and cancellation so a timed-out embed cannot leave the lock held and block later searches. **`BRAINLAYER_EMBED_TIMEOUT_MS`** now rejects non-finite or non-positive values (defaults to 1000 ms). Empty no-hit responses include **`search_mode`** and **`fallback_reason`**, with user-facing text explaining FTS-only fallback when embedding was skipped.
> 
> The **prompt hook** applies the same pattern via `HYBRID_IN_FLIGHT` around timed hybrid search so overlapping hook invocations do not spawn parallel hybrid workers. Tests cover parallel suppression, lock release edge cases, and timeout env validation.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 93caf0ea60c8ca483f8e54e4f32648477e815bf0. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Fall back to FTS when embedding is busy or times out in search handler
> - Adds a module-level `_EMBED_IN_FLIGHT` lock in [`search_handler.py`](https://github.com/EtanHey/brainlayer/pull/524/files#diff-1f952ebc3cd8256a9bc05612e63df4c5cdcc59336a777da97375fa88fdb8bac0) so only one embedding can run at a time; concurrent requests immediately fall back to FTS-only with `fallback_reason='embed_busy'`.
> - On `asyncio.TimeoutError`, the handler falls back to FTS-only with `fallback_reason='embed_timeout'`; on general exceptions, falls back with `fallback_reason='embed_error:<ExceptionClass>'`.
> - Fixes `BRAINLAYER_EMBED_TIMEOUT_MS` validation to reject NaN/Inf and non-positive values, defaulting to 1000ms instead of using invalid values.
> - Structured and empty search results now include `search_mode` and `fallback_reason` metadata; empty-result text reflects FTS-only fallback when applicable.
> - Risk: embedding is now fully serialized — parallel search requests that previously ran concurrent embeddings will now fall back to FTS instead.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 93caf0e.</sup>
> 
>

## Release Notes

* **New Features**
  * Search operations now enforce configurable timeouts for embedding queries, improving responsiveness.
  * Automatic fallback to full-text search when embedding times out or encounters errors.
  * Search results now indicate fallback mode and reason when embedding is skipped.

* **Tests**
  * Added comprehensive test coverage for embedding timeout and fallback scenarios.

### CODE REVIEW

File: src/brainlayer/mcp/search_handler.py [hunk 1]
```diff
@@ -1783,40 +1786,83 @@ async def _search(
...
+                finally:
+                    release_embed_lock_once()
+
+            try:
+                try:
+                    embed_future = loop.run_in_executor(None, embed_query_guarded)
+                except Exception:
+                    release_embed_lock_once()
+                    raise
+                query_embedding = await asyncio.wait_for(embed_future, timeout=embed_timeout_ms / 1000.0)
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/mcp/search_handler.py:1839]: This is addressed in the current `93caf0e` diff: `_search` now has an explicit `except asyncio.CancelledError:` branch immediately after the `wait_for`, calls `release_if_embed_never_started()`, then re-raises. The regression `test_brain_search_releases_embed_lock_when_cancelled_before_executor_starts` reproduces exactly the queued-future cancellation case and failed before this change. Full pre-push passed after the fix: 3025 passed, 9 skipped, 61 deselected, 1 xfailed, plus MCP registration, isolated eval/hook routing, bun, and FTS determinism.
